counsel.el

```diff
@@ -2245,18 +2245,50 @@ regex string."
 (ivy-set-occur 'counsel-ag 'counsel-ag-occur)
 (ivy-set-display-transformer 'counsel-ag 'counsel-git-grep-transformer)
 
+(defconst user-input-switch "-- ")
+(defconst user-input-switch-length (length user-input-switch))
+
+(defun parse-user-input (string)
+  "Parse user input in STRING into a pair of two strings: (query extra-ag-args).
+Where query is the search term and extra-ag-args are any additional switches."
```

Comment:
"Parse", "user input", and "query" are too vague, and "extra-ag-args" is too specific. What this function is actually doing is splitting a command into its constitutent program and switches (see [`(elisp) Tips for Defining`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Tips-for-Defining.html)). Note also that the Elisp convention is to write metasyntactic variables like `query` and `extra-ag-args` in uppercase; see [`(elisp) Documentation Tips`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Documentation-Tips.html).

As such, I propose something more along the lines of:

```el
(defun counsel-split-command (command)
  "Split shell COMMAND into its program and switches.
Return pair of corresponding strings (PROGRAM . SWITCHES)."
```

Reply:
Good points and thanks for including the links.

The function is not separating the command from the switches tho:
It operates only on the *arguments* to a command (i.e.: `ag`) but the command itself is not part of the input. It splits this argument into a pair of *switches* and the *search-term* (or *search-query* or *search-regexp*).

The current order of the resulting pair probably leads to confusion:
Now it is `(query . extra-ag-args)`, but it could be more clear if it followed the order in which the user would write them: `(extra-ag-args . query)` or based on your feedback: `(switches . search-term)`

How about?

```elisp
(defun counsel-split-command-args (arguments)
  "Split ARGUMENTS into its switches and search term.
Return pair of corresponding strings (SWITCHES. SEARCH-TERM)."
```

Reply:
SGTM. My only remaining bikesheds would be to use the prefix `counsel--` instead of `counsel-`, as this is an internal function (see the previous link for Tips for Defining), and to suggest "needle" as a shorter alternative to "search-term".

Reply:
Done